app/src/org/commcare/utils/EncryptionUtils.java

```diff
@@ -1,17 +1,274 @@
+    private static final String BLOCK_MODE = KeyProperties.BLOCK_MODE_CBC;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String PADDING = KeyProperties.ENCRYPTION_PADDING_PKCS7;
+    private static KeyStore keystoreSingleton = null;
+
+    private static KeyStore getKeystore() throws KeyStoreException, CertificateException,
+            IOException, NoSuchAlgorithmException {
+        if (keystoreSingleton == null) {
+            keystoreSingleton = KeyStore.getInstance(KEYSTORE_NAME);
+            keystoreSingleton.load(null);
+        }
+
+        return keystoreSingleton;
+    }
+
+    //Gets the SecretKey from the Android KeyStore (creates a new one the first time)
+    private static Key getKey(Context context, boolean trueForEncrypt)
+            throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException,
+            UnrecoverableEntryException, InvalidAlgorithmParameterException, NoSuchProviderException {
+        KeyStore keystore = getKeystore();
+
+        if(doesKeystoreContainEncryptionKey()) {
+            KeyStore.Entry existingKey = keystore.getEntry(SECRET_NAME, null);
+            if (existingKey instanceof KeyStore.SecretKeyEntry entry) {
+                return entry.getSecretKey();
+            }
+            if (existingKey instanceof KeyStore.PrivateKeyEntry entry) {
+                return trueForEncrypt ? entry.getCertificate().getPublicKey() : entry.getPrivateKey();
+            }
+            else { return null; }
+        } else {
+            return generateKeyInKeystore(context, trueForEncrypt);
+        }
+    }
+
+    private static boolean doesKeystoreContainEncryptionKey() throws CertificateException,
+            KeyStoreException, IOException, NoSuchAlgorithmException {
+        KeyStore keystore = getKeystore();
+
+        return  keystore.containsAlias(SECRET_NAME);
+    }
+
+    private static Key generateKeyInKeystore(Context context, boolean trueForEncrypt) throws NoSuchAlgorithmException, NoSuchProviderException, InvalidAlgorithmParameterException {
+        if(android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
+            KeyGenerator keyGenerator = KeyGenerator.getInstance(
+                    KeyProperties.KEY_ALGORITHM_AES, KEYSTORE_NAME);
+            KeyGenParameterSpec keySpec = new KeyGenParameterSpec.Builder(SECRET_NAME,
+                    KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
+                    .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+                    //.setUserAuthenticationValidityDurationSeconds(AUTH_VALID_DURATION_IN_SECOND)
+                    .build();
+
+            keyGenerator.init(keySpec);
+            return keyGenerator.generateKey();
+        }
+        else {
+            KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA", KEYSTORE_NAME);
```

Comment:
The [SO post](https://stackoverflow.com/a/50389146/3811963) actually suggests to not use this key pair directly for encryption but use it to encrypt another AES key that we generate and actually use for encrypting the data. That way we can always use AES encryption to encrypt data and the use of RSA key pair is just for making the key secure to be able to store it on the device.

Reply:
I could be mistaken (still getting comfortable with cryptography), but I think the current code is as suggested in that SO post. The post says (for API < 23)  we should use an RSA key pair to encrypt a random AES key, with the RSA key stored in the KeyStore and the encrypted AES key stored somewhere easily accessible. Then later we can retrieve the RSA key, decrypt the AES key, and finally access our DB.

In our implementation, SQLCipher handles the AES encryption. We use ConnectIDDatabaseHelper.getConnectDBPassphrase() to handle retrieving (or creating the first time) the AES key. The first time accessing the DB, we create the AES key, encrypt it, encode it to a base64 string, and store it in the CommCare Global DB. Diving into the code that encrypts the AES key: EncryptionUtils.getKey() looks for an existing key in the KeyStore, and otherwise calls generateInKeystore(). This is the function that generates the key in the KeyStore directly (for API >= 23) or manually generates an RSA key pair to store (for earlier APIs).

By the time we attempt to access the encrypted Connect DB, the RSA key has already been generated and stored in the KeyStore, and the AES key has been generated, encrypted, encoded, and stored in the global DB. So now it's just a matter of retrieving the RSA key, decrypting the AES key, and passing it to the Connect DB.

Reply:
Thanks for the detailed explanation. That makes sense to me clears things up.

Although I think there is a fault in this whole mechanism that we are depening on the OS version to decide the Encryption algorithm [here](https://github.com/dimagi/commcare-android/blob/dv/connect_id/app/src/org/commcare/utils/EncryptionUtils.java#L174). This can lock an user out from their data after doing an OS update. We somehow need to know the `transformation` based on the `key` stored in keystore itself to avoid this.

Reply:
bumping

Reply:
Excellent point, the DB would definitely break when a user upgrades from API <31 to >=31.

I changed the code so that the key retrieval and generation code (i.e. getKey()) now also returns the transformation string (via a new KeyAndTransform wrapper class). Now when generating the key, we can also create the correct transformation depending on the API level. When retrieving the key from the Keystore, we can tell which encryption was used during DB initialization depending on whether the retrieved key is a SecretKeyEntry (for AES) or PrivateKeyEntry (for RSA), and create the appropriate transformation string.

A caveat of this solution is that we'll continue using RSA encryption after the user upgrades their OS, but it seems ideally we would want to migrate from RSA to AES to take advantage of the increased security. Is this something we should explore further?